Generalizes disabled link style as a utility class

anselmbradford opened the pull request:
Moves disabled link styles to utility class so that list_link can jump_link can share a disabled link class.

## Additions

- Added `u-link__disabled` utility class for styling disabled links.

## Removals

- Removed `link_list__disabled` class.

## Testing

- Any disabled links (e.g. the homepage), should have a not-allowed cursor and grey color.

## Review

- @jimmynotjim 
- @sebworks 
- @KimberlyMunoz 

## Notes

- Spun off from DF-2099.

Scotchester commented:
Nice :+1:

Scotchester commented:
P.S.: I stealth-updated your description to correct the class name you originally entered as `.u-list__disabled`

Scotchester commented:
Upon further consideration: I don't think we have a precedent for using the modifier syntax in a utility class. I think they tend to ignore BEM convention somewhat. Additionally, `link` does not exist as a block itself, so it doesn't seem right to modify it.

How about `.u-disabled-link`?

jimmynotjim commented:
Looks good and agree with @Scotchester on the `link` block. Doesn't need to exist since this is just a utility with a single purpose.

anselmbradford commented:
> P.S.: I stealth-updated your description to correct the class name you originally entered as .u-list__disabled

Thanks! :smile: Realized I've totally gargled the changelog with transposing these values, will fix.

> Upon further consideration: I don't think we have a precedent for using the modifier syntax in a utility class. I think they tend to ignore BEM convention somewhat. Additionally, link does not exist as a block itself, so it doesn't seem right to modify it.

Hm, good point. Though this may be an opportunity to create a precedent. 
Do you think this is most appropriate as a utility class?

For buttons we have `btn__disabled` and for links we have `list_link__disabled` and `jump-link__disabled` (present in another branch). To just deal with links and not lists, say we had another block called `skip-link`, with a modifier `skip-link__disabled`. If we wanted to generalize a disabled link class for both those link blocks, it seems helpful to have the name of the class follow the format of the blocks they are intended to operate on, so:
 ```html
<a class="link jump-link link__disabled">
<a class="link skip-link link__disabled">
```
I didn't particularly want to create an empty `link` block though to make the component semantically complete, so came to this, which didn't need a block but still implied it operated on one:

```html
<a class="jump-link u-link__disabled">`.
<a class="skip-link u-link__disabled">`.
```

I'd argue the `u-` prefix for utility classes would mitigate confusion on whether they were coming from a true block or meant to operate on one.

Anyway though, that's just some background. I'm not wedded to it if it still seems to make sense to have it `u-disabled-link`.

sebworks commented:
:+1:

anselmbradford commented:
Pushed through https://github.com/cfpb/cfgov-refresh/commit/d86bd1aebf3d1af98f598c0b4a9566841a64e651 with `u-disabled-link`, but let me know if there are any further thoughts on conventions around the utility classes :smiley_cat:

Scotchester commented on src/static/css/footer.less:
```diff
@@ -240,7 +240,7 @@
         }
     }
 
-    .list_link__disabled {
```
Why is this still here? Why are and of the disabled-link things in this file still here?

anselmbradford replied:
Mm, good catch. This is an artifact of replacing the existing styles. I'll need to check that the links don't need to be a different color here.

As an aside I think this should go back to `u-link__disabled`, at least in the interim. I noticed `footer.less` is filled with `u-link__` patterns, and found it's actually an [established pattern in cf-core](https://github.com/cfpb/cf-core/blob/gh-pages/src/cf-utilities.less#L393-L487).

Scotchester replied:
Hmm... good point. I forgot about those. In retrospect, I don't love naming utility mixins that way, but don't think we want to go changing that now.

anselmbradford commented:
Okay, updated!
- Changed everything to `u-link__disabled`.
- Removed footer and header specific disabled link styles. **Note: The header was setting a lighter grey, but I couldn't see where it was in use and it doesn't seem to make sense to have two disabled link styles with slightly different depths of grey.**

Can I get another review please?

himedlooff commented:
Utilities aren't states. For example `. expandable__expanded ` and `.btn__disabled`.

anselmbradford commented:
> Utilities aren't states. For example . expandable__expanded and .btn__disabled.

That makes sense, though we do have existing states that follow this pattern, for example [the hover states](https://github.com/cfpb/cf-core/blob/gh-pages/src/cf-utilities.less#L478-L487). It seems like maybe we need a `link` component (in cf-typography?) that would usurp all the `u-link__*` utility classes and mixins. For this PR I'd err on the side of fitting the existing patterns so everything could be refactored together. What do you think?

himedlooff commented:
Yeah `u-link` is a confusing one but they're all mixins no?

anselmbradford commented:
> Yeah u-link is a confusing one but they're all mixins no?

Ahh yes, that's true! I overlooked the parentheses on the ones that didn't have parameters. What do you think for this PR? `u-disabled-link`?

himedlooff commented:
I like your idea of doing what's been done `.u-link__disabled` so it can be consistently changed if ever refactored, but I defer to you all.

KimberlyMunoz commented:
:+1:

jimmynotjim commented:
:+1:
